beacon-chain/rpc/eth/debug/handlers.go

```diff
@@ -0,0 +1,147 @@
+package debug
+
+import (
+	"context"
+	"encoding/json"
+	"net/http"
+
+	"github.com/gorilla/mux"
+	"github.com/prysmaticlabs/prysm/v4/api"
+	"github.com/prysmaticlabs/prysm/v4/beacon-chain/rpc/eth/helpers"
+	"github.com/prysmaticlabs/prysm/v4/beacon-chain/rpc/eth/shared"
+	http2 "github.com/prysmaticlabs/prysm/v4/network/http"
+	"github.com/prysmaticlabs/prysm/v4/runtime/version"
+	"go.opencensus.io/trace"
+)
+
+// GetBeaconStateSSZ returns the SSZ-serialized version of the full beacon state object for given state ID.
+//
+// DEPRECATED: please use GetBeaconStateV2 instead
+func (s *Server) GetBeaconStateSSZ(w http.ResponseWriter, r *http.Request) {
+	ctx, span := trace.StartSpan(r.Context(), "debug.GetBeaconStateSSZ")
+	defer span.End()
+
+	stateId := mux.Vars(r)["state_id"]
+	if stateId == "" {
+		http2.HandleError(w, "state_id is required in URL params", http.StatusBadRequest)
+		return
+	}
+	st, err := s.Stater.State(ctx, []byte(stateId))
+	if err != nil {
+		shared.WriteStateFetchError(w, err)
+		return
+	}
+	sszState, err := st.MarshalSSZ()
+	if err != nil {
+		http2.HandleError(w, "Could not marshal state into SSZ: "+err.Error(), http.StatusInternalServerError)
+		return
+	}
+	http2.WriteSsz(w, sszState, "beacon_state.ssz")
+}
+
+// GetBeaconStateV2 returns the full beacon state for a given state ID.
+func (s *Server) GetBeaconStateV2(w http.ResponseWriter, r *http.Request) {
+	ctx, span := trace.StartSpan(r.Context(), "debug.GetBeaconStateV2")
+	defer span.End()
+
+	stateId := mux.Vars(r)["state_id"]
+	if stateId == "" {
+		http2.HandleError(w, "state_id is required in URL params", http.StatusBadRequest)
+		return
+	}
+
+	if http2.SszRequested(r) {
+		s.getBeaconStateSSZV2(ctx, w, []byte(stateId))
+	} else {
+		s.getBeaconStateV2(ctx, w, []byte(stateId))
+	}
+}
+
+// getBeaconStateV2 returns the JSON-serialized version of the full beacon state object for given state ID.
```

Comment:
i guess more appropriate is it sets the response writer and not return anything.

Reply:
but it makes me wonder for readability, would it be better to have these functions return the errors and not write the errors in these functions, instead at the end of the `func (s *Server) GetBeaconStateV2(w http.ResponseWriter, r *http.Request)` call a generic helper that either writes the json response or the error and have this function only worry about the business logic and creating the response

Reply:
> i guess more appropriate is it sets the response writer and not return anything

The meaning of "returns" here is that it returns the value in an HTTP response. I think it should be clear from the context, given it's an HTTP handler. 

> but it makes me wonder for readability

Is the current code unreadable?

Reply:
I suppose you can easily set the error for the body but not return the response in the way it is now. you and I are fully aware of this, but perhaps as time goes on newer teammates might not be familiar with this pattern.